tcc に RV32 の対象 (riscv32) を足す。

上流: https://github.com/TinyCC/tinycc  mob 2ba12e83b3599ca8f5d50c179fe5138fe956f0c9
説明: docs/stage015-riscv32.md

RISC-V の backend (riscv64-gen.c / riscv64-link.c) は RV32 と RV64 で
大半が共通なので，ファイルを写さず**同じファイルを XLEN で切り替える**
形にしてある。この patch の差分がそのまま「RV32 に要る違い」である。

対象は -DTCC_TARGET_RISCV64 -DTCC_TARGET_RISCV32 の 2 つを定義して選ぶ。
TCC_TARGET_RISCV64 は「RISC-V であること」を表し，TCC_TARGET_RISCV32 が
「語長が 32 bit であること」を絞り込む。こうすると core (tccgen.c・
tccelf.c・tccasm.c・tccdbg.c) にある 40 箇所あまりの `#ifdef
TCC_TARGET_RISCV64` を 1 つも書き換えずに済む。

diff -u -r --unidirectional-new-file -x .git -x '*.o' -x '*.a' -x config.h -x config.mak -x tcc tcc/Makefile ../../tmp/tcc/src/Makefile
--- a/Makefile
+++ b/Makefile
@@ -117,6 +117,7 @@
 DEF-arm64-OpenBSD  = $(DEF-arm64) -DTARGETOS_OpenBSD
 DEF-arm64-win32    = $(DEF-arm64) -DTCC_TARGET_PE
 DEF-riscv64        = -DTCC_TARGET_RISCV64
+DEF-riscv32        = -DTCC_TARGET_RISCV64 -DTCC_TARGET_RISCV32
 DEF-c67            = -DTCC_TARGET_C67 -w # disable warnigs
 DEF-x86_64-FreeBSD = $(DEF-x86_64) -DTARGETOS_FreeBSD
 DEF-x86_64-NetBSD  = $(DEF-x86_64) -DTARGETOS_NetBSD
@@ -134,7 +135,7 @@
 
 # cross compiler targets to build
 TCC_X = i386 x86_64 i386-win32 x86_64-win32 x86_64-osx arm arm64 arm64-win32 arm-wince c67
-TCC_X += riscv64 arm64-osx
+TCC_X += riscv64 arm64-osx riscv32
 # TCC_X += arm-fpa arm-fpa-ld arm-vfp arm-eabi
 
 # cross libtcc1.a targets to build
@@ -217,6 +218,7 @@
 arm64-win32_FILES = $(arm64_FILES) tccpe.c
 c67_FILES = $(CORE_FILES) c67-gen.c c67-link.c tcccoff.c
 riscv64_FILES = $(CORE_FILES) riscv64-gen.c riscv64-link.c riscv64-asm.c
+riscv32_FILES = $(riscv64_FILES)
 
 TCCDEFS_H$(subst yes,,$(CONFIG_predefs)) = tccdefs_.h
 
diff -u -r --unidirectional-new-file -x .git -x '*.o' -x '*.a' -x config.h -x config.mak -x tcc tcc/lib/Makefile ../../tmp/tcc/src/lib/Makefile
--- a/lib/Makefile
+++ b/lib/Makefile
@@ -74,6 +74,8 @@
 OBJ-arm-eabihf = $(OBJ-arm)
 OBJ-arm-wince = $(ARM_O) $(WIN_O)
 OBJ-riscv64 = $(RISCV64_O) armflush.o $(LIN_O)
+RISCV32_O = libtcc1.o $(COMMON_O)
+OBJ-riscv32 = $(RISCV32_O) armflush.o $(LIN_O)
 
 OBJ-extra = $(filter $(EXTRA_O),$(OBJ-$T))
 OBJ-libtcc1 = $(addprefix $(X),$(filter-out $(OBJ-extra),$(OBJ-$T)))
diff -u -r --unidirectional-new-file -x .git -x '*.o' -x '*.a' -x config.h -x config.mak -x tcc tcc/riscv64-gen.c ../../tmp/tcc/src/riscv64-gen.c
--- a/riscv64-gen.c
+++ b/riscv64-gen.c
@@ -21,12 +21,21 @@
 #define REG_IRE2 (TREG_R(1)) // int 2nd return register number
 #define REG_FRET (TREG_F(0)) // float return register number
 
+#ifdef TCC_TARGET_RISCV32
+#define PTR_SIZE 4
+
+#define LDOUBLE_SIZE 8
+#define LDOUBLE_ALIGN 8
+
+#define MAX_ALIGN 8
+#else
 #define PTR_SIZE 8
 
 #define LDOUBLE_SIZE 16
 #define LDOUBLE_ALIGN 16
 
 #define MAX_ALIGN 16
+#endif
 
 #define CHAR_IS_UNSIGNED
 
@@ -46,7 +55,11 @@
 
 ST_DATA const char * const target_machine_defs =
     "__riscv\0"
+#ifdef TCC_TARGET_RISCV32
+    "__riscv_xlen 32\0"
+#else
     "__riscv_xlen 64\0"
+#endif
     "__riscv_flen 64\0"
     "__riscv_div\0"
     "__riscv_mul\0"
@@ -55,7 +68,7 @@
     "__riscv_float_abi_double\0"
     ;
 
-#define XLEN 8
+#define XLEN PTR_SIZE
 
 #define TREG_RA 17
 #define TREG_SP 18
@@ -1172,10 +1185,13 @@
     gen_opil(op, 0);
 }
 
+#ifndef TCC_TARGET_RISCV32
+/* 32 bit 対象の 64 bit 演算は tccgen.c の gen_opl が組み立てる */
 ST_FUNC void gen_opl(int op)
 {
     gen_opil(op, 1);
 }
+#endif
 
 ST_FUNC void gen_opf(int op)
 {
diff -u -r --unidirectional-new-file -x .git -x '*.o' -x '*.a' -x config.h -x config.mak -x tcc tcc/tcc.h ../../tmp/tcc/src/tcc.h
--- a/tcc.h
+++ b/tcc.h
@@ -401,6 +401,14 @@
 # define ElfW_Rel ElfW(Rela)
 # define SHT_RELX SHT_RELA
 # define REL_SECTION_FMT ".rela%s"
+#elif defined TCC_TARGET_RISCV32
+/* RISC-V の再配置は 32 bit でも加数を項目に持つ (RELA) */
+# define ELFCLASSW ELFCLASS32
+# define ElfW(type) Elf##32##_##type
+# define ELFW(type) ELF##32##_##type
+# define ElfW_Rel ElfW(Rela)
+# define SHT_RELX SHT_RELA
+# define REL_SECTION_FMT ".rela%s"
 #else
 # define ELFCLASSW ELFCLASS32
 # define ElfW(type) Elf##32##_##type
@@ -1720,7 +1728,9 @@
 
 /* ------------ riscv64-gen.c ------------ */
 #ifdef TCC_TARGET_RISCV64
-ST_FUNC void gen_opl(int op);
+#ifndef TCC_TARGET_RISCV32
+ST_FUNC void gen_opl(int op); /* 32 bit 対象では tccgen.c が持つ */
+#endif
 ST_FUNC void gen_va_start(void);
 ST_FUNC void arch_transfer_ret_regs(int);
 ST_FUNC void gen_cvt_sxtw(void);
